Pyrado/pyrado/algorithms/meta/sprl.py

```diff
@@ -0,0 +1,576 @@
+                previous_distribution.distribution, distribution.distribution
+            )
+            return kl_divergence.detach().numpy()
+
+        def kl_constraint_fn_prime(x):
+            distribution = previous_distribution.from_stacked(x)
+            kl_divergence = to.distributions.kl_divergence(
+                previous_distribution.distribution, distribution.distribution
+            )
+            grads = to.autograd.grad(kl_divergence, distribution.parameters())
+            return np.concatenate([g.detach().numpy() for g in grads])
+
+        kl_constraint = NonlinearConstraint(
+            fun=kl_constraint_fn,
+            lb=-np.inf,
+            ub=self._kl_constraints_ub,
+            jac=kl_constraint_fn_prime,
+            # keep_feasible=True,
+        )
+
+        def performance_constraint_fn(x):
+            distribution = previous_distribution.from_stacked(x)
+            performance = self._compute_expected_performance(distribution, contexts, contexts_old_log_prob, values)
+            return performance.detach().numpy()
+
+        def performance_constraint_fn_prime(x):
+            distribution = previous_distribution.from_stacked(x)
+            performance = self._compute_expected_performance(distribution, contexts, contexts_old_log_prob, values)
+            grads = to.autograd.grad(performance, distribution.parameters())
+            return np.concatenate([g.detach().numpy() for g in grads])
+
+        performance_contraint = NonlinearConstraint(
+            fun=performance_constraint_fn,
+            lb=self._performance_lower_bound,
+            ub=np.inf,
+            jac=performance_constraint_fn_prime,
+            # keep_feasible=True,
+        )
+
+        # Optionally clip the bounds of the new variance
+        bounds = None
+        x0, _, x0_cov_indices = previous_distribution.get_stacked(return_mean_cov_indices=True)
+        if self._kl_threshold != -np.inf and (self._kl_threshold < kl_divergence):
+            lower_bound = np.ones_like(x0) * -np.inf
+            if x0_cov_indices is not None:
+                lower_bound[x0_cov_indices] = self._std_lower_bound
+            upper_bound = np.ones_like(x0) * np.inf
+            # bounds = Bounds(lb=lower_bound, ub=upper_bound, keep_feasible=True)
+            bounds = Bounds(lb=lower_bound, ub=upper_bound)
+            x0 = np.clip(x0, lower_bound, upper_bound)
+
+        objective_fn: Optional[Callable[..., Tuple[np.array, np.array]]] = None
+        result = None
+        constraints = None
+
+        # Check whether we are already above our performance threshold
+        if performance_constraint_fn(x0) >= self._performance_lower_bound:
+            self._performance_lower_bound_reached = True
+            constraints = [kl_constraint, performance_contraint]
+
+            # We now optimize based on the kl-divergence between target and context distribution by minimizing it
+            def objective(x):
```

Comment:
Please add a one line doc string """..."""

Reply:
Done in 681e167a4c9b3d5be2c3d560c9f145c2842a87c1